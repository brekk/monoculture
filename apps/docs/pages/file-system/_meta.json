{
  "packages-file-system-fs.js": "fs"
}