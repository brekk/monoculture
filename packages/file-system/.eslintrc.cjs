module.exports = {
  root: true,
  extends: ['monoculture'],
  ignorePatterns: ['file-system.js'],
}
