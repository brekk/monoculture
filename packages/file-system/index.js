export * from './find-up'
export * from './exec'
export * from './fs'
export * from './interpret'
export * from './path'
