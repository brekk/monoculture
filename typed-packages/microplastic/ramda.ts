// so remeda doesn't seem like it actually solves much at first blush

import { curry } from 'ramda'

interface Transformer<F, T> {
  (x: F): T
}

interface EqualishBy<F, T> {
  (t: Transformer<F, T>, expected: T, x: F): boolean
  (t: Transformer<F, T>, expected: T): (x: F) => boolean
  (t: Transformer<F, T>): (expected: T, x: F) => boolean
}

export function equalishBy<F, T>(a: Transformer<F, T>, b: T, c: F) {
  const raw: EqualishBy<F, T> = curry((transform, expected, x) => {
    const t = transform(x)
    return t === expected
  })
  return raw(a, b, c)
}
