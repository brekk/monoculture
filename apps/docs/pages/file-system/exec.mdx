# exec

Call external processes and consume them as a Future-wrapped value. This is a light wrapper around `execa`, so please see [execa](https://github.com/sindresorhus/execa)'s documentation for more info.



## execWithConfig

Consume external commands as a Future-wrapped value.

### Usage
```js
import { execWithConfig } from 'file-system'
import { fork } from 'fluture'
fork(console.warn)(console.log)(
  execWithConfig(
    function customCancellationFunction() {},
    'echo',
    { cleanup: true },
    ['ahoy']
  )
)
```

#### See also
 - execWithCancel
 - exec
## execWithCancel

Consume external commands as a Future-wrapped value.

### Usage
```js
import { execWithCancel } from 'file-system'
import { fork } from 'fluture'
fork(console.warn)(console.log)(
  execWithCancel(
    function customCancellationFunction() {},
    'echo',
    ['ahoy']
  )
)
```

#### See also
 - execWithConfig
 - exec
## exec

Consume external commands as a Future-wrapped value.

### Usage
```js
import { exec } from 'file-system'
import { fork } from 'fluture'
fork(console.warn)(console.log)(
  exec(
    'echo',
    ['ahoy']
  )
)
```

#### See also
 - execWithConfig
 - execWithCancel