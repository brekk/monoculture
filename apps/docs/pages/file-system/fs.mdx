# fs

A Future-wrapped `fs` API, for future-based, lazy, easy-to-model asynchrony that makes it easy to manipulate the file system.



## localize

make a file string relative

### Usage
```js
import { localize } from 'file-system'
console.log(`support ${localize('business')}`)
// support ./business
```

## readFileWithFormatAndCancel

Read a file asynchronously as a Future-wrapped value.

### Usage
```js
import { fork } from 'fluture'
import { readFile } from 'file-system'
fork(console.warn)(console.log)(
  readFileWithFormatAndCancel(() => process.exit(), 'utf8', './README.md')
)
```

#### See also
 - readFileWithCancel
 - readFile
## readFileWithCancel

Read a file asynchronously as a Future-wrapped value.

### Usage
```js
import { fork } from 'fluture'
import { readFile } from 'file-system'
fork(console.warn)(console.log)(
  readFileWithCancel(() => process.exit(), './README.md')
)
```

#### See also
 - readFileWithFormatAndCancel
 - readFile
## readFile

Read a file asynchronously as a Future-wrapped value.

### Usage
```js
import { fork } from 'fluture'
import { readFile } from 'file-system'
fork(console.warn)(console.log)(readFile('./README.md'))
```

#### See also
 - readFileWithFormatAndCancel
 - readFileWithCancel

## readJSONFileWithCancel

Read a JSON file asynchronously as a Future-wrapped value, given a cancellation function

### Usage
```js
import { fork } from 'fluture'
import { readJSONFile } from 'file-system'
fork(console.warn)(console.log)(readJSONFile('./package.json'))
```

#### See also
 - readFile
 - readJSONFile

## readJSONFile

Read a JSON file asynchronously as a Future-wrapped value

### Usage
```js
import { fork } from 'fluture'
import { readJSONFile } from 'file-system'
fork(console.warn)(console.log)(readJSONFile('./package.json'))
```

#### See also
 - readFile
 - readJSONFileWithCancel

## readDirWithConfigAndCancel

Read a glob asynchronously as a Future-wrapped value,
with configuration and a cancellation function.
Configuration is passed to [glob](https://www.npmjs.com/package/glob)

### Usage
```js
import { fork } from 'fluture'
import { readDirWithConfigAndCancel } from 'file-system'
// [...]
pipe(
  fork(console.warn)(console.log)
)(readDirWithConfigAndCancel(cancellationFn, { ignore: ['node_modules/**'] }, 'src/*'))
```

#### See also
 - readDirWithConfig
 - readDir

## readDirWithConfig

Read a glob asynchronously as a Future-wrapped value, with configuration.
Configuration is passed to [glob](https://www.npmjs.com/package/glob)

### Usage
```js
import { fork } from 'fluture'
import { readDirWithConfig } from 'file-system'
// [...]
pipe(
  fork(console.warn)(console.log)
)(readDirWithConfig({ ignore: ['node_modules/**'] }, 'src/*'))
```

#### See also
 - readDirWithConfigAndCancel
 - readDir

## readDir

Read a glob asynchronously as a Future-wrapped value, default config assumed.
Configuration is passed to [glob](https://www.npmjs.com/package/glob)

### Usage
```js
import { fork } from 'fluture'
import { readDir } from 'file-system'
// [...]
fork(console.warn)(console.log)(readDir('src/*'))
```

#### See also
 - readDirWithConfigAndCancel
 - readDirWithConfig

## writeFileWithConfigAndCancel

Write to a file, with configuration and a cancellation function.
Unlike `fs.writeFile`, this will return the written value as a Future-wrapped value.

### Usage
```js
import { fork } from 'fluture'
import { writeFileWithConfig } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  writeFileWithConfigAndCancel(
    cancellationFunction,
    { ...fs.writeFileConfig },
    'my-file.txt',
    'hey I am a file'
  )
)
```

#### See also
 - writeFileWithConfig
 - writeFile

## writeFileWithConfig

Write to a file, with configuration.
Unlike `fs.writeFile`, this will return the written value as a Future-wrapped value.

### Usage
```js
import { fork } from 'fluture'
import { writeFileWithConfig } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  writeFileWithConfig(
    { encoding: 'utf8' },
    'my-file.txt',
    'hey I am a file'
  )
)
```

#### See also
 - writeFileWithConfigAndCancel
 - writeFile

## writeFile

Write to a file, assuming `'utf8'`.
Unlike `fs.writeFile`, this will return the written value as a Future-wrapped value.

### Usage
```js
import { fork } from 'fluture'
import { writeFileWithConfig } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  writeFileWithConfig(
    'my-file.txt',
    'hey I am a file'
  )
)
```

#### See also
 - writeFileWithConfigAndCancel
 - writeFileWithConfig

## removeFileWithConfigAndCancel

Remove a file, configurably, with cancellation.
Unlike `fs.rm`, this returns the path of the deleted file as a Future-wrapped string.

### Usage
```js
import { fork } from 'fluture'
import { removeFileWithConfigAndCancel } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  removeFileWithConfigAndCancel(
    cancellationFn,
    { ...fs.removeFileConfig },
    'my-file.txt'
  )
)
```

#### See also
 - removeFileWithConfig
 - removeFile

## removeFileWithConfig

Remove a file, configurably.
Unlike `fs.rm`, this returns the path of the deleted file as a Future-wrapped string.

### Usage
```js
import { fork } from 'fluture'
import { removeFileWithConfig } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  removeFileWithConfig(
    { ...fs.removeFileConfig },
    'my-file.txt'
  )
)
```

#### See also
 - removeFileWithConfigAndCancel
 - removeFile

## removeFile

Remove a file.
Unlike `fs.rm`, this returns the path of the deleted file as a Future-wrapped string.

### Usage
```js
import { fork } from 'fluture'
import { removeFile } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  removeFile(
    'my-file.txt'
  )
)
```

#### See also
 - removeFileWithConfigAndCancel
 - removeFileWithConfig

## removeFilesWithConfigAndCancel

Remove multiple files, configurably, with a cancellation function.

### Usage
```js
import { fork } from 'fluture'
import { removeFilesWithConfig } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  removeFilesWithConfigAndCancel(
    cancellationFn,
    { parallel: 30 },
    [...list, ...of, ...thirty, ...files]
  )
)
```

#### See also
 - removeFilesWithConfig
 - removeFiles

## removeFilesWithConfig

Remove multiple files, configurably.

### Usage
```js
import { fork } from 'fluture'
import { DEFAULT_REMOVAL_CONFIG, removeFilesWithConfig } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  removeFilesWithConfig(
    DEFAULT_REMOVAL_CONFIG,
    [...list, ...of, ...thirty, ...files]
  )
)
```

#### See also
 - removeFilesWithConfigAndCancel
 - removeFiles

## removeFilesWithConfig

Remove multiple files, configurably.

### Usage
```js
import { fork } from 'fluture'
import { removeFilesWithConfig } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  removeFilesWithConfig(
    [...list, ...of, ...thirty, ...files]
  )
)
```

#### See also
 - removeFilesWithConfigAndCancel
 - removeFiles

## mkdirWithCancel

Make a directory, given a cancellation function.
Returns a Future-wrapped file path as a discrete value upon success.

### Usage
```js
import { fork } from 'fluture'
import { mkdirWithCancel } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  mkdirWithCancel(
    () => {},
    {},
    'my-dir'
  )
)
```

#### See also
 - mkdir
 - mkdirp

## mkdir

Make a directory
Returns a Future-wrapped file path as a discrete value upon success.

### Usage
```js
import { fork } from 'fluture'
import { mkdir } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  mkdir(
    {},
    'my-dir'
  )
)
```

#### See also
 - mkdirWithCancel
 - mkdirp

## mkdirp

Make a directory, recursively.
Returns a Future-wrapped file path as a discrete value upon success.

### Usage
```js
import { fork } from 'fluture'
import { mkdir } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  mkdir(
    {},
    'my-dir'
  )
)
```

#### See also
 - mkdirWithCancel
 - mkdir

## writeFileWithAutoPath

Write a file to a nested folder and automatically create needed folders, akin to `mkdir -p`

### Usage
```js
import { fork } from 'fluture'
import { writeFileWithAutoPath } from 'file-system'
// [...]
fork(console.warn)(console.log)(
  writeFileWithAutoPath(
    "folders/you/must/exist/file.biz",
    "my cool content"
  )
)
```