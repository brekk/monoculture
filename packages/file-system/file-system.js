#!/usr/bin/env node

// src/flexeca.js
import { pipe, curry, ifElse, propOr } from "ramda";
import { execa } from "execa";
import { Future } from "fluture";
var didFail = propOr(true, "failed");
var fail = propOr("Something broke", "stderr");
var flexecaWithCanceller = curry(
  (cancellation, cmd, args) => new Future((bad, good) => {
    execa(cmd, args).catch(pipe(fail, (z) => `FAILED: "${z}"`, bad)).then(ifElse(didFail, pipe(fail, bad), good));
    return cancellation;
  })
);
var flexeca = flexecaWithCanceller(() => {
});

// src/fs.js
import {
  rm as __rm,
  mkdir as __mkdir,
  readFile as __readFile,
  writeFile as __writeFile,
  access as __access,
  constants
} from "node:fs";
import { propOr as propOr2, without, curry as curry2, pipe as pipe2, map } from "ramda";
import { chain, chainRej, Future as Future2, parallel } from "fluture";
import glob from "glob";
var localize = (z) => `./${z}`;
var readFile = (x) => new Future2((bad, good) => {
  __readFile(x, "utf8", (err, data) => err ? bad(err) : good(data));
  return () => {
  };
});
var readJSONFile = pipe2(readFile, map(JSON.parse));
var readDirWithConfig = curry2(
  (conf, g) => Future2((bad, good) => {
    try {
      glob(
        g,
        conf,
        (e, x) => (
          // thus far I cannot seem to ever call `bad` from within here
          e ? bad(e) : good(x)
        )
      );
    } catch (e) {
      bad(e);
    }
    return () => {
    };
  })
);
var readDir = readDirWithConfig({});
var writeFileWithConfig = curry2(
  (conf, file, content) => new Future2((bad, good) => {
    __writeFile(file, content, conf, (e) => {
      if (e) {
        bad(e);
        return;
      }
      good(content);
    });
    return () => {
    };
  })
);
var writeFile = writeFileWithConfig({ encoding: "utf8" });
var removeFileWithConfig = curry2(
  (options, fd) => new Future2((bad, good) => {
    __rm(fd, options, (err) => err ? bad(err) : good(fd));
    return () => {
    };
  })
);
var DEFAULT_REMOVAL_CONFIG = {
  force: false,
  maxRetries: 0,
  recursive: false,
  retryDelay: 100,
  parallel: 10
};
var removeFile = removeFileWithConfig(DEFAULT_REMOVAL_CONFIG);
var removeFilesWithConfig = curry2(
  (conf, list) => pipe2(
    map(removeFileWithConfig(without(["parallel"], conf))),
    parallel(propOr2(10, "parallel", conf))
  )(list)
);
var removeFiles = removeFilesWithConfig(DEFAULT_REMOVAL_CONFIG);
var mkdir = curry2(
  (conf, x) => new Future2((bad, good) => {
    __mkdir(x, conf, (err) => err ? bad(err) : good(x));
    return () => {
    };
  })
);
var mkdirp = mkdir({ recursive: true });
var access = curry2(
  (permissions, filePath) => new Future2((bad, good) => {
    __access(filePath, permissions, (err) => err ? bad(err) : good(true));
    return () => {
    };
  })
);
var exists = access(constants.F_OK);
var directoryOnly = (filePath) => filePath.slice(0, filePath.lastIndexOf("/"));
var writeFileWithAutoPath = curry2(
  (filePath, content) => pipe2(
    directoryOnly,
    (dir) => pipe2(
      exists,
      chainRej(() => mkdirp(dir))
    )(dir),
    chain(() => writeFile(filePath, content))
  )(filePath)
);
var rm = curry2(
  (conf, x) => new Future2((bad, good) => {
    __rm(x, conf, (err) => err ? bad(err) : good(x));
    return () => {
    };
  })
);
var rimraf = rm({ force: true, recursive: true });

// src/path.js
import { join, normalize } from "node:path";
import { curry as curry3 } from "ramda";
var pathRelativeTo = curry3((pwd, x) => {
  if (typeof pwd !== "string" || typeof x !== "string") {
    throw new Error("Cannot normalize bad paths.");
  }
  return join(pwd, normalize(x));
});

// src/interpret.js
import { Future as Future3 } from "fluture";
var interpret = (filepath) => Future3((bad, good) => {
  import(filepath).catch(bad).then(good);
  return () => {
  };
});
export {
  DEFAULT_REMOVAL_CONFIG,
  access,
  directoryOnly,
  exists,
  flexeca,
  flexecaWithCanceller,
  interpret,
  localize,
  mkdir,
  mkdirp,
  pathRelativeTo,
  readDir,
  readDirWithConfig,
  readFile,
  readJSONFile,
  removeFile,
  removeFileWithConfig,
  removeFiles,
  removeFilesWithConfig,
  rimraf,
  rm,
  writeFile,
  writeFileWithAutoPath,
  writeFileWithConfig
};
