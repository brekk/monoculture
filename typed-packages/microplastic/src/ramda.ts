import { curry } from 'ramda'

interface Transformer<F, T> {
  (arg: F): T
}

export interface EqualishBy<F, T> {
  (transform: Transformer<F, T>, expected: T, x: F): boolean
  (transform: Transformer<F, T>, expected: T): (x: F) => boolean
  (transform: Transformer<F, T>): (expected: T, x: F) => boolean
}

export const equalishBy = curry(
  <F, T>(transform: Transformer<F, T>, expected: T, x: F) =>
    transform(x) === expected
)
