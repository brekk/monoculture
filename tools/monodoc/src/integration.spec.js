// import { cwd } from 'node:process'
import { join, identity as I, pipe, map, propOr } from 'ramda'
import { fork } from 'fluture'
import { flexeca } from 'file-system'

const run = pipe(flexeca('./monodoc-cli.js'), map(propOr('', 'stdout')))

const goodrun = (args, expectation = I) =>
  test(`monodoc ${join(' ')(args)}`, done => {
    pipe(
      run,
      fork(done)(z => {
        expectation(z)
        expect(z).toBeTruthy()
        expect(z).toMatchSnapshot()
        done()
      })
    )(args)
  })

goodrun([`fixture.ts`])
