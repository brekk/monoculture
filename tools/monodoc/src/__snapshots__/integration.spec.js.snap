// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monodoc fixture.ts 1`] = `
" monodoc 

magical documentation tooling âœ¨

  -i / --input
  	A file to read!

  -o / --output
  	The file to output!

  -s / --search
  	The glob to use for searching (default: '**//*.{js,jsx,ts,tsx}')

  -a / --artifact
  	Would you like to create an artifact file?
  	(Useful for downstream transformation)

  -g / --ignore
  	Files to ignore when searching, can be specified multiple times

  -k / --color
  	Render stuff in color

  -h / --help
  	This text!"
`;
